erxupdate/ntopnghosts.sh

```diff
@@ -0,0 +1,429 @@
+		fi
+	fi
+else 
+	cp /etc/nprobe/hostnamesubstitute /tmp/hostnamesubstitute
+fi
+#main script follows
+/opt/vyatta/bin/vyatta-op-cmd-wrapper show dhcp leases | tail -n +3| tr -s " " | sed 's/\^I//' | sort >/tmp/ntopng_newdhcp
+#cp tmp/ntopng_newdhcp /tmp/ntopng_newdhcp # for testing only
+#cp tmp/ntopng_olddhcp /tmp/ntopng_olddhcp # for testing only
+if [[ ! -f "/tmp/ntopng_olddhcp" ]]; then
+	if [[ "$TELEGRAMENABLE" == "YES" ]]; then
+		MESSAGE="`uname -n` has recently rebooted.  There may be some repeat notifications for devices seen in the last 24 hours"
+		TELEGRAMHOSTNAME="`uname -n`"
+		echo `sendmessage`
+	fi
+	# we initialise olddhcp with the contents of newdhcp so that the script doesn't get bogged down with processing every device it can see - that would
+	# take longer than 1 minute and therefore would end up with multiple copies of the script running after a reboot.
+	cp /tmp/ntopng_newdhcp /tmp/ntopng_olddhcp
+fi
+comm -3 /tmp/ntopng_olddhcp /tmp/ntopng_newdhcp |grep $SUBNET >/tmp/ntopng_dhcpcomm
+# we want to show any non-printable characters including ^I (tab)
+cat -vet /tmp/ntopng_dhcpcomm >/tmp/ntopng_dhcpdiff
+rm /tmp/ntopng_dhcpcomm
+# next we need to remove ^I (tab) characters if they are present
+sed -i 's/\^I//' /tmp/ntopng_dhcpdiff
+# because we are using cat -vet to remove non-printable characters including ^I (tab) we get $ on the end of lines, so we remove that on the next line
+# so that they don't end up in the hostnames
+sed -i 's/\$$//' /tmp/ntopng_dhcpdiff
+#cat /tmp/ntopng_dhcpdiff
+if [[ "$DEBUG" != "" ]]; then 
+	trap - DEBUG
+	trap
+fi
+while IFS= read -r line; do
+    LASTHOSTNAME=$HOSTNAME
+	LASTMAC=$MAC
+	#echo $line
+	# we will process this file line by line to check each HOSTNAME
+    # echo "Text read from file: $line"
+    IP=`echo "$line" | cut -d ' ' -f 1`
+    HOSTNAME=`echo "$line" | cut -d ' ' -f 6`
+	if [[ "$HOSTNAME" == "?" || "$HOSTNAME" == "" ]]; then
+	    HOSTNAME="unknown"
+	fi
+	# Note that while NTOPNG doesn't require the MAC address we are including it in the hostname as hostname(macaddress) as this helps with uniqueness (iPhones and iPads
+	# for example all default to iPhone or iPad, unlike Android that uses a unique name).  We also use hostname and mac address for device name substitutions and for 
+	# reporting to telegram if you have enabled that functionality.
+	MAC=`echo "$line" | cut -d ' ' -f 2`
+    echo "Checking IP=$IP HOSTNAME=$HOSTNAME MAC=$MAC"
+	# echo "line=$line"
+	if [[ "$HOSTNAME $MAC" != "$LASTHOSTNAME $LASTMAC" ]]; then
+    	NTOPNGHOSTNAME="$HOSTNAME($MAC)"
+        # check if we need to replace the hostname with something more meaningful that we have on file
+		if [[ -f "/tmp/hostnamesubstitute" ]]; then
+            SUB=`grep -i "^$HOSTNAME $MAC" /tmp/hostnamesubstitute | cut -d ' ' -f 3-100`
+            if [[ "$SUB" != "" ]]; then
+                echo "Replacing $HOSTNAME with $SUB"
+                NTOPNGHOSTNAME="$SUB"
```

Comment:
SUB can contain injection chars if it comes from an untrusted source, causing arbitrary shell code execution.

Reply:
SUB doesn't come from an untrusted source.  SUB is derived from a line in /tmp/hostnamesubstitute.  Hostnamesubstitute is either sourced from /etc/nprobe/ which is safe within the router (unless someone hacks the router - in which case the system is fully compromised already) or from a webserver controlled by the person operating the router - they define the URL that they trust to download this from and they define the contents of the file on the webserver.  If they don't secure the webserver and the webserver gets hacked then there isn't anything we can do to protect them as they already have a big problem.